pkg/controllers/build/application_controller_test.go

```diff
@@ -0,0 +1,389 @@
+/*
+Copyright 2019 the original author or authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package build_test
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/projectriff/system/pkg/apis"
+	buildv1alpha1 "github.com/projectriff/system/pkg/apis/build/v1alpha1"
+	kpackbuildv1alpha1 "github.com/projectriff/system/pkg/apis/thirdparty/kpack/build/v1alpha1"
+	"github.com/projectriff/system/pkg/controllers/build"
+	rtesting "github.com/projectriff/system/pkg/controllers/testing"
+	"github.com/projectriff/system/pkg/refs"
+	"github.com/projectriff/system/pkg/tracker"
+)
+
+func TestApplicationReconcile(t *testing.T) {
+	testNamespace := "test-namespace"
+	testName := "test-application"
+	testKey := types.NamespacedName{Namespace: testNamespace, Name: testName}
+	testImagePrefix := "example.com/repo"
+	testGitUrl := "git@example.com:repo.git"
+	testGitRevision := "master"
+	testSha256 := "cf8b4c69d5460f88530e1c80b8856a70801f31c50b191c8413043ba9b160a43e"
+	testConditionReason := "TestReason"
+	testConditionMessage := "meaningful, yet concise"
+
+	scheme := runtime.NewScheme()
+	_ = clientgoscheme.AddToScheme(scheme)
+	_ = kpackbuildv1alpha1.AddToScheme(scheme)
+	_ = buildv1alpha1.AddToScheme(scheme)
+
+	table := rtesting.Table{{
+		Name: "application does not exist",
+		Key:  testKey,
+	}, {
+		Name: "create kpack image, default image",
+		Key:  testKey,
+		GivenObjects: []runtime.Object{
+			&corev1.ConfigMap{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "riff-build",
+					Namespace: testNamespace,
+				},
+				Data: map[string]string{
+					"default-image-prefix": testImagePrefix,
+				},
+			},
+			&buildv1alpha1.Application{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      testName,
+					Namespace: testNamespace,
+				},
+				Spec: buildv1alpha1.ApplicationSpec{
+					Source: &buildv1alpha1.Source{
+						Git: &buildv1alpha1.Git{
+							URL:      testGitUrl,
+							Revision: testGitRevision,
+						},
+					},
+				},
+			},
+		},
+		ExpectCreates: []runtime.Object{
+			&kpackbuildv1alpha1.Image{
```

Comment:
Inline declarations will end up being unmaintainable. As the tests are fleshed out to cover error percolation and other scenarios, we'll end up with a lot of duplication which will require parallel maintenance.

The main issue with adding helper functions is the potential for their names clashing with other similar functions in the same package. One solution would be to define helper function variables in the scope of the test function. A downside of this is that they need to be defined before being used and so need to appear at the top. Another solution would be to use longer names and a naming convention to avoid clashes. A more radical solution would be to move each controller to a separate package.